partcad/src/partcad/actions/part_actions.py

```
@@ -0,0 +1,142 @@
+        pc_logging.info(f"Ad-hoc conversion successful: {converted_path}")
+
+    # Define target path inside the project
+    target_path = (Path(project.path) / f"{name}.{kind}").resolve()
+
+    if target_path.exists() and source_path.samefile(target_path):
+        pc_logging.warning(f"Skipping copy: source and target paths are the same ({source_path}).")
+    else:
+        try:
+            shutil.copy2(source_path, target_path)
+        except shutil.Error as e:
+            raise ValueError(f"Failed to copy '{source_path}' → '{target_path}': {e}")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)